### PR TITLE
rgw: experimental support for restoring a lost bucket index

### PR DESCRIPTION
**Pull request by ivancich**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/rgw_admin.cc [hunk 1]
```diff
@@ -340,6 +340,7 @@ void usage()
   cout << "   --bucket=<bucket>         Specify the bucket name. Also used by the quota command.\n";
   cout << "   --pool=<pool>             Specify the pool name. Also used to scan for leaked rados objects.\n";
   cout << "   --object=<object>         object name\n";
+  cout << "   --objects-file=<file>     file containing a list of object names to process\n";
```
Comment (mattbenjamin) [inline src/rgw/rgw_admin.cc:343]: this seems sound and can handle our use case.  I can hazily imagine a world where you give a pool name and maybe a pattern string that would try to restore all matching objects?

File: src/rgw/rgw-restore-bucket-index [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+# quotes
+get_field_q() {
+  grep "\"$2\":" $1 | sed "s/.*\".*\": \"\(.*\)\".*/\1/"
+}
+
+bucketentrymeta=/tmp/bucketentrymeta.$$
+bucketinstmeta=/tmp/bucketinstmeta.$$
+objects=/tmp/objects.$$
+
```
Comment (mattbenjamin) [inline src/rgw/rgw-restore-bucket-index:33]: i see, the script handles the groveling

File: src/rgw/rgw-restore-bucket-index [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+get_pool() {
+  if [ -n "$1" ] ;then
+    echo "$1"
+  else
+    radosgw-admin zone get >$data_pools 2>/dev/null
+    pool_count=$(grep "\"data_pool\":" $data_pools | wc -l)
+    if [ "$pool_count" -ne 1 ] ;then
+      echo "Error: unable to determine data pool; please specify on command-line."
+      usage
```
Comment (cbodley) [inline src/rgw/rgw-restore-bucket-index:43]: if you want to get clever about this part, you can determine the correct pool name from the bucket instance metadata's "placement_rule" field

that string will either in the form `"{placement id}/{storage class}"`, or just the placement id if the storage class is the default "STANDARD" one. you'd use those keys to look up the correct data_pool under the zone's placement_pools:
```
    "placement_pools": [
        {
            "key": "default-placement",
            "val": {
                "index_pool": "default.rgw.buckets.index",
                "storage_classes": {
                    "STANDARD": {
                        "data_pool": "default.rgw.buckets.data"
                    }
                },
                "data_extra_pool": "default.rgw.buckets.non-ec",
                "index_type": 0,
                "inline_data": true
            }
```

File: src/rgw/rgw-restore-bucket-index [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+fi
+
+echo num_shards is $num_shards
+
+if [ "$num_shards" -le 0 ] ;then
+   echo "Error: number of shards -- $num_shards -- is invalid"
+   exit 3
+fi
+
+(rados -p $pool ls | grep "^${marker}_" | sed "s/^${marker}_\(.*\)/\1/" >$obj_list ) 2>/dev/null
```
Comment (cbodley) [inline src/rgw/rgw-restore-bucket-index:94]: isn't this going to pick up 'namespaced' rados objects like multipart parts and shadow/tail objects?

**Comment by ivancich** [comment]
@linuxkidd : I'm looking for holes in the bash script, as in what could go wrong? A later version can be polished up. But if you see anything feel free to add a review.

File: src/rgw/rgw-restore-bucket-index (hunk 4 above)
Comment (ivancich) [reply src/rgw/rgw-restore-bucket-index:122]: Good catch; need to prohibit the next underscore.

File: src/rgw/rgw-restore-bucket-index (hunk 3 above)
Comment (ivancich) [reply src/rgw/rgw-restore-bucket-index:75]: That kind of logic would seem to exceed `sed`'s ability to parse. We'd have to force `jq` to be installed, which Matt thinks is reasonable.

Comment (ivancich) [reply src/rgw/rgw-restore-bucket-index]: I think I've added the appropriate logic to `get_pool()` to cover explicit pools and placement rules. Would you mind taking a look, @cbodley ?

Comment (ivancich) [reply src/rgw/rgw-restore-bucket-index]: (using jq)

Comment (cbodley) [reply src/rgw/rgw-restore-bucket-index]: insofar as i can decipher the awk and jq, it looks like it's doing the right thing

**Comment by ivancich** [comment]
@cbodley and @mattbenjamin : Would you mind re-reviewing this?

File: src/rgw/driver/rados/rgw_rados.cc [hunk 5]
```diff
@@ -3592,6 +3592,21 @@ int RGWRados::rewrite_obj(RGWBucketInfo& dest_bucket_info, const rgw_obj& obj, c
                        attrset, 0, real_time(), NULL, dpp, y);
 }
 
+int RGWRados::reindex_obj(RGWBucketInfo& dest_bucket_info, const rgw_obj& obj, const DoutPrefixProvider *dpp, optional_yield y)
+{
+  Bucket target(this, dest_bucket_info);
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_rados.cc:3597]: probably a good place to return an error like `ENOTSUP` if `dest_bucket_info.versioned()`

File: src/rgw/rgw_admin.cc [hunk 6]
```diff
@@ -170,7 +171,8 @@ void usage()
   cout << "  object put                 put object\n";
   cout << "  object stat                stat an object for its metadata\n";
   cout << "  object unlink              unlink object from bucket index\n";
-  cout << "  object rewrite             rewrite the specified object\n";
+  cout << "  object rewrite             rewrite the object(s) indicated by --bucket and either --object or --objects-file\n";
+  cout << "  object reindex             reindex the object(s) indicated by --bucket and either --object or --objects-file\n";
```
Comment (cbodley) [inline src/rgw/rgw_admin.cc:175]: will need to update src/test/cli/radosgw-admin/help.t accordingly

File: src/rgw/rgw-restore-bucket-index [hunk 7]
```diff
@@ -0,0 +1,182 @@
+#!/usr/bin/env bash
+
+# version 2023-03-03
```
Comment (cbodley) [inline src/rgw/rgw-restore-bucket-index:3]: as a repair tool, i think this deserves a concise description of what it does and when you'd want to use it

File: src/rgw/rgw-restore-bucket-index [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+# examine number of bucket index shards
+num_shards=$(jq ".data.bucket_info.num_shards" $bkt_inst)
+echo number of bucket index shards is $num_shards
+if [ "$num_shards" -le 0 ] ;then
+   echo "Warning: number of shards ($num_shards) should be non-zero; fixing...."
+
+   sed 's/"num_shards": 0/"num_shards": 1/' $bkt_inst >$bkt_inst_new
+   radosgw-admin metadata put bucket.instance:${bucket}:${bucket_id} --infile=$bkt_inst_new
+   radosgw-admin metadata get bucket.instance:${bucket}:${bucket_id}
```
Comment (cbodley) [inline src/rgw/rgw-restore-bucket-index:134]: i don't think section is necessary after https://github.com/ceph/ceph/pull/50329

if the bucket does still have an index object, this change would render it inaccessible (see https://github.com/ceph/ceph/pull/50329#issuecomment-1450855002). we'd go on to create new a shard object, but leak the original

if the intent is to use this script on a cluster that doesn't have that fix, i guess it's fine to include this section in a hotfix, but we shouldn't merge it to main/reef. if we're delivering a hotfix that includes your changes to rgw_admin.cc, we might as well include https://github.com/ceph/ceph/pull/50329 too?

**Review by cbodley — COMMENTED** [review]
maybe drop the `--objects-file` changes to `object rewrite`? not sure it's worth maintaining it there

File: src/test/cli/radosgw-admin/help.t [hunk 9]
```diff
@@ -42,6 +42,7 @@
     object stat                stat an object for its metadata
     object unlink              unlink object from bucket index
     object rewrite             rewrite the specified object
+    object reindex             reindex the object(s) indicated by --bucket and either --object or --objects-file
```
Comment (cbodley) [inline src/test/cli/radosgw-admin/help.t:45]: the --objects-file and --infile changes to radosgw-admin usage will also need to be reflected here

**Review by cbodley — APPROVED** [review]
other than the test.t failures, this looks great. if we plan to make it not-experimential in the future, we'll want to write some tests for it

**Comment by ivancich** [comment]
Teuthology Results:
* https://pulpito.ceph.com/ivancich-2023-03-09_00:46:24-rgw-wip-restore-bucket-index-distro-default-smithi/
* https://pulpito.ceph.com/ivancich-2023-03-09_00:46:52-rgw-wip-restore-bucket-index-distro-default-smithi/

File: src/rgw/rgw-restore-bucket-index [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+    # warn user and get permission to proceed
+    echo "NOTICE: This tool is currently considered EXPERIMENTAL."
+    echo "The list of objects that we will attempt to restore can be found in \"$obj_list\"."
+    echo "Please review the object names in that file (either below or in another window/terminal) before proceeding."
+    while true ; do
+	read -p "Type \"proceed!\" to proceed, \"view\" to view object list, or \"q\" to quit: " action
+	if [ "$action" == "q" ] ;then
+	    echo "Exiting..."
+	    clean
+	    exit 0
```
Comment (linuxkidd) [inline src/rgw/rgw-restore-bucket-index:176]: Is there a reason to use `super_exit` here, vs `exit 0` ?

**Review by linuxkidd — COMMENTED** [review]
I don't see any show stoppers in the script.  Just a few suggestions / queries.

File: src/rgw/rgw-restore-bucket-index (hunk 10 above)
Comment (ivancich) [reply src/rgw/rgw-restore-bucket-index:176]: When you're in a function `exit` takes you out of the function but puts you back in the caller, where `super_exit` will end the entire script.

File: src/rgw/rgw-restore-bucket-index [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+# bucket instance data.
+get_pool() {
+  # command-line
+  if [ -n "$1" ] ;then
+    echo "$1"
+    exit 0
+  fi
+
+  # explicit_placement
+  expl_pool=$(strip_quotes $(jq '.data.bucket_info.bucket.explicit_placement.data_pool' $bkt_inst))
```
[the remark this answers is not in the document]
Comment (ivancich) [reply src/rgw/rgw-restore-bucket-index:88]: Thanks for looking, @linuxkidd ! I'll look into raw output.

File: ceph.spec.in [hunk 12]
```diff
@@ -2132,6 +2132,7 @@ fi
 %{_bindir}/rgw-gap-list-comparator
 %{_bindir}/rgw-orphan-list
 %{_bindir}/rgw-policy-check
+%{_bindir}/rgw-restore-bucket-index
```
Comment (cbodley) [inline ceph.spec.in:2135]: QE is expecting these scripts to be installed along with `radosgw-admin`, which is packaged in the `%files common` section. the `radosgw` package is generally only installed on the radosgw server nodes

Comment (ivancich) [reply ceph.spec.in:2135]: Thanks, @cbodley. As you know (but documenting for the PR here), there's an add-on PR to address this: https://github.com/ceph/ceph/pull/50545.

**Pull request merged**